### DISCUSSION TITLE
Asset Library: Partial and case-insensitive searching

### DISCUSSION THREAD

**Discussion by aaronatbissell** [aws/aws-connected-device-framework#48] [Q&A]
Our team is looking for ways to do both partial searches and case-insensitive searching. Is this possible with Asset Library/Neptune? If not - is it on the roadmap?

I noticed the assetlibrary-client has some functions like `contains` and `startsWith` and `endsWith`. These seem like good candidates for partial searches, but they don't appear to be functioning that way. Can you point us to some documentation for these search params?

**Comment by boardthatpowder** [comment]
@aaronatbissell since the asset library was implemented, there have been 2 Neptune related feature updates which can help with this: 

- Tinkerpop 3.4.1 text predicates such as `containing` - https://aws.amazon.com/blogs/database/amazon-neptune-now-supports-tinkerpop-3-4-features/
- Full-text search features via Elastic Search - https://docs.aws.amazon.com/neptune/latest/userguide/full-text-search-query-execution.html

Note that the Tinkerpop text predicates are still case-sensitive, therefore if you decide to implement these then one way of allowing these to be case-insensitive is to have a lower cased version of your existing fields you wish to search added to your device/groups, then search using a lower cased version of the search term.

We do have work in progress to add both options listed above, but may be 1-2 months before being released.

**Comment by boardthatpowder** [comment]
@aaronatbissell I have just been corrected. The Tinkerpop 3.4.1 text predicates were added to the assetlibrary in https://github.com/aws/aws-connected-device-framework/releases/tag/1.0.3.
